### PR TITLE
Feature.add additional email

### LINKED ISSUE #4756: Add support for adding additional emails
This issue will be used to track content-server related tasks for https://github.com/mozilla/fxa-features/issues/27

The UX and flows are still in flux but at a high level should include

- [ ] Should only be able to add another email with verified session
- [ ] Show primary email and additional emails in one section
- [ ] Ability to add additional email
- [ ] Ability to remove additional email
- [ ] Can not change primary
- [ ] Mail check
- [ ] Email bounced notifications on UI

### PR DESCRIPTION
**Pull request by vladikoff**
Fixes #4756

Reopening for docker stuff **feature.add-additional-email** branch name / tag

## TODO

* [x] finish up functional tests
  opened: https://github.com/mozilla/fxa-content-server/issues/5054
* [x] DONE - gating on the content-server for emails (`@mozilla, @softvision, @restmail`)
* [x] unit tests
* [x] remove ` - echo "Tests totally passed ;)"` in circle.yml

## DONE

* [x] “Add” button should be blue in settings
* [x] The Secondary email section should always be open if the secondary email is not verified
* [x] The section should say “cancel” instead of done
* [x] Emails have wrong footer compared to the mock up from Ryan: 
   1. `undefined` in email: in the `verify secondary email` template we have `<a undefined>please change your password.</a>`
   1.  FOR NOW: leave that in, DO NOT Follow the Invision design (do not add "intruders")
   1.  FOR LATER: discuss unsubscribe (the flow confusing for that). 

* [x] Customs configs on docker builds-> We need to disable customs on emails3.dev.lcip.org
* [x] IP_PROFILE need to disable it on emails3.dev.lcip.org
* [x] input placeholder is wrong in the secondary email section
* [x] ~~In the email: please change password link doesn’t link. check footer of the email in the design. That link is not there.~~ See Above
* [x] Always enable “Refresh” button
* [x] **Delete account didn’t delete the secondary email**
* [x] DONE button should be grey if email is waiting to be verified.
* [x] **Should reset password -> send to secondary?**
  1. You should get an error for now, you have to put your PRIMARY EMAIL in the field.
* [x] remind Ryan F: "need to make an error for users trying to register a **with** secondary email" // review all new errors with Ryan F.

* MINOR: verification link says “Damaged” when verifying after some other user took the email... Should it say "Expired" ?

### CODE REVIEW

**Comment by vladikoff** [comment]
Ref: https://github.com/mozilla/fxa-dev/blob/docker/aws/environments/emails.yml#L13

**Comment by vladikoff** [comment]
@vbudhram If I try to `Reset Account` then I get Unknown Account (even if I have that email verified as secondary email):

**Comment by vladikoff** [comment]
The Update Branch button got itself into conflicts. I had to rebase this branch from scratch and remove merge commits. let's not use the "Update branch" button at least in this branch for now. It creates bad merge commits that later cannot be rebased.

File: app/scripts/models/email.js [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+ */
+
+define(function (require, exports, module) {
+  'use strict';
+
+  const Backbone = require('backbone');
+
+  var Email = Backbone.Model.extend({
+    defaults: {
+      email: undefined,
```
Comment (vladikoff) [inline app/scripts/models/email.js]: this needs to use `null` like other files in `models/`

File: app/scripts/views/decorators/progress_indicator.js [hunk 2]
```diff
@@ -18,8 +18,16 @@ define(function (require, exports, module) {
   const p = require('lib/promise');
   const ProgressIndicator = require('views/progress_indicator');
 
-  function showProgressIndicator(handler, _el) {
+  // Return a promise delayed by ms
+  function delay(ms) {
+    var deferred = p.defer();
+    setTimeout(deferred.resolve, ms);
```
Comment (vladikoff) [inline app/scripts/views/decorators/progress_indicator.js]: other files use `this.setTimeout` or use the timer-mixin. we need to add that

File: app/scripts/views/mixins/settings-panel-mixin.js [hunk 3]
```diff
@@ -89,12 +89,15 @@ define(function (require, exports, module) {
       $('.settings-unit').removeClass('open');
     },
 
-    displaySuccess (msg) {
+    displaySuccess (msg, options = { closePanel: true}) {
```
Comment (vladikoff) [inline app/scripts/views/mixins/settings-panel-mixin.js]: nit : style near `clonePanel`

File: app/scripts/views/ready.js [hunk 4]
```diff
@@ -49,7 +53,7 @@ define(function (require, exports, module) {
       headerId: 'fxa-sign-up-complete-header',
       headerTitle: t('Account verified'),
       readyToSyncText: t('You are now ready to use %(serviceName)s')
-    }
+    },
```
Comment (vladikoff) [inline app/scripts/views/ready.js]: need to remove dangling comma

File: app/scripts/views/settings.js [hunk 5]
```diff
@@ -30,6 +31,7 @@ define(function (require, exports, module) {
   const Template = require('stache!templates/settings');
 
   var PANEL_VIEWS = [
+    EmailsView,
```
Comment (vladikoff) [inline app/scripts/views/settings.js]: nit: needs to be alphabetical

File: app/scripts/views/settings/emails.js [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+    refresh: showProgressIndicator(function() {
+      return this.render();
+    }, EMAIL_REFRESH_SELECTOR, EMAIL_REFRESH_DELAYMS),
+
+    resend (event) {
+      const email = $(event.currentTarget).data('id');
+      const account = this.getSignedInAccount();
+      return account.resendEmailCode(email)
+        .then(() => {
+          this.displaySuccess(t('Verification emailed to ') + email, {
```
Comment (vladikoff) [inline app/scripts/views/settings/emails.js]: This is not gonna work, this needs to string interpolate the translate string

File: app/scripts/views/settings/emails.js [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+        });
+    },
+
+    submit () {
+      const newEmail = this.getElementValue('input.new-email');
+      if (this.isPanelOpen() && newEmail) {
+        const account = this.getSignedInAccount();
+        return account.recoveryEmailCreate(newEmail)
+          .then(() => {
+            this.displaySuccess(t('Verification emailed to ') + newEmail, {
```
Comment (vladikoff) [inline app/scripts/views/settings/emails.js]: This is not gonna work, this needs to string interpolate the translate string

**Comment by vladikoff** [comment]
* [ ] “Add” button should be blue in settings
* [ ] The Secondary email section should always be open if the secondary email is not verified
* [ ] The section should say “cancel” instead of done
* [ ] Emails have wrong footer compared to the mock up from Ryan
* [ ] Customs configs on docker builds-> We need to disable customs on emails3.dev.lcip.org
* [ ] IP_PROFILE need to disable it on emails3.dev.lcip.org
* [ ] input placeholder is wrong in the secondary email section
* [ ] In the email: please change password link doesn’t link. check footer of the email in the design. That link is not there.
* [ ] Always enable “Refresh” button
* [ ] **Delete account didn’t delete the secondary email**
* [ ] DONE button should be grey if email is waiting to be verified.
* [ ] **Should reset password -> send to secondary?**
* [ ] MINOR: verification link says “Damaged” when verifying after some other user took the email... Should it say "Expired" ?

File: app/scripts/lib/auth-errors.js [hunk 8]
```diff
@@ -145,6 +145,48 @@ define(function (require, exports, module) {
       errno: 135,
       message: t('Unable to deliver email')
     },
+    // Secondary Email errors
+    EMAIL_EXISTS: {
+      errno: 136,
+      message: t('This email was already verified by another user')
+    },
+    EMAIL_DELETE_PRIMARY: {
```
Comment (vladikoff) [inline app/scripts/lib/auth-errors.js]: from mtg: we probably do not need this one

File: app/scripts/lib/auth-errors.js [hunk 9]
```diff
@@ -145,6 +145,48 @@ define(function (require, exports, module) {
...
+      errno: 137,
+      message: t('Can not delete primary email')
+    },
+    SESSION_UNVERIFIED: {
+      errno: 138,
+      message: t('Unable to perform action, unverified session')
```
Comment (vladikoff) [inline app/scripts/lib/auth-errors.js]: also might not need this in the content-server, just like https://github.com/mozilla/fxa-content-server/pull/4996/files#r115769238 . At least we can remove the translation bit.

File: app/scripts/lib/auth-errors.js [hunk 10]
```diff
@@ -145,6 +145,48 @@ define(function (require, exports, module) {
...
+      errno: 141,
+      message: t('Account already exists')
+    },
+    LOGIN_WITH_SECONDARY_EMAIL: {
+      errno: 142,
+      message: t('Primary account email required for sign-in')
+    },
+    SECONDARY_EMAIL_UNKNOWN: {
+      errno: 143,
+      message: t('Cannot use this email to sign in')
```
Comment (vladikoff) [inline app/scripts/lib/auth-errors.js]: we need to check this one, is this user facing?  from mtg: it should probably say `'Primary account email required for sign-in'`

File: app/scripts/lib/auth-errors.js [hunk 11]
```diff
@@ -145,6 +145,48 @@ define(function (require, exports, module) {
...
+    },
+    VERIFIED_SECONDARY_EMAIL_EXISTS: {
+      errno: 144,
+      message: t('Address in use by another account')
+    },
+    RESET_PASSWORD_WITH_SECONDARY_EMAIL: {
+      errno: 145,
+      message: t('Reset with this email type is not currently supported. Use the primary email of this account')
```
Comment (vladikoff) [inline app/scripts/lib/auth-errors.js]: need to change: 'Primary account email required for reset'

File: app/scripts/lib/auth-errors.js (hunk 10 above)
Comment (vbudhram) [inline app/scripts/lib/auth-errors.js]: Nope, looks like it is not user facing. Will remove 👍

File: app/scripts/lib/auth-errors.js [hunk 12]
```diff
@@ -145,6 +145,36 @@ define(function (require, exports, module) {
...
+      errno: 136,
+      message: t('This email was already verified by another user')
+    },
+    EMAIL_PRIMARY_EXISTS: {
+      errno: 139,
+      message: t('Secondary email must be different than your account email')
+    },
+    EMAIL_VERIFIED_PRIMARY_EXISTS: {
+      errno: 140,
+      message: t('Account already exists')
```
Comment (shane-tomlinson) [inline app/scripts/lib/auth-errors.js]: Should this and the next error have different strings so we can differentiate in Sentry?

File: app/scripts/lib/auth-errors.js [hunk 13]
```diff
@@ -145,6 +145,36 @@ define(function (require, exports, module) {
...
+      errno: 141,
+      message: t('Account already exists')
+    },
+    LOGIN_WITH_SECONDARY_EMAIL: {
+      errno: 142,
+      message: t('Primary account email required for sign-in')
+    },
+    VERIFIED_SECONDARY_EMAIL_EXISTS: {
+      errno: 144,
+      message: t('Address in use by another account')
```
Comment (shane-tomlinson) [inline app/scripts/lib/auth-errors.js]: What's the difference between this and # 136?

File: app/scripts/templates/settings/emails.mustache [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+            </p>
+            {{#hasSecondaryEmail}}
+                <ul class="email-list button-row">
+                    {{#emails}}
+                        {{^isPrimary}}
+                            <li class="email-address">
+                                <div class="address">{{email}}</div>
+                                <div class="details">
+                                    {{#verified}}
+                                        <div class="verified"> {{#t}}verified{{/t}}</div>
```
Comment (shane-tomlinson) [inline app/scripts/templates/settings/emails.mustache]: Do you want the extra space in front of the `{`?

File: app/scripts/views/settings/emails.js [hunk 15]
```diff
@@ -0,0 +1,156 @@
...
+
+    resend (event) {
+      const email = $(event.currentTarget).data('id');
+      const account = this.getSignedInAccount();
+      return account.resendEmailCode(email)
+        .then(() => {
+          this.displaySuccess(Strings.interpolate(t('Verification emailed to %(email)s'), { email: email }), {
+            closePanel: false
+          });
+          this.render();
```
Comment (shane-tomlinson) [inline app/scripts/views/settings/emails.js]: Do you need to render if you are immediately navigating?

File: app/scripts/views/settings/emails.js [hunk 16]
```diff
@@ -0,0 +1,156 @@
...
+    refresh: showProgressIndicator(function() {
+      return this.render();
+    }, EMAIL_REFRESH_SELECTOR, EMAIL_REFRESH_DELAYMS),
+
+    resend (event) {
+      const email = $(event.currentTarget).data('id');
+      const account = this.getSignedInAccount();
+      return account.resendEmailCode(email)
+        .then(() => {
+          this.displaySuccess(Strings.interpolate(t('Verification emailed to %(email)s'), { email: email }), {
```
Comment (shane-tomlinson) [inline app/scripts/views/settings/emails.js]: I personally find this string slightly confusing.

File: app/scripts/lib/auth-errors.js (hunk 12 above)
Comment (vladikoff) [inline app/scripts/lib/auth-errors.js]: Sentry has the errno AFAIK

File: app/scripts/views/settings/emails.js (hunk 16 above)
Comment (vladikoff) [inline app/scripts/views/settings/emails.js]: @shane-tomlinson what do you mean? the wording?

Comment (shane-tomlinson) [inline app/scripts/views/settings/emails.js]: @shane-tomlinson what do you mean? the wording?

`Verification emailed to...` 

When sending a signup email, we say:

> A verification link has been sent&nbsp;to %(escapedEmail)s

**Pull request merged**